File: lib/blockdb.py
"""Database abstraction for the blockchain and related structures on disk."""

import json
import os
import struct
from typing import Dict, List, TYPE_CHECKING

import plyvel

# pylint: disable=C0411
from .block import Block
from .crypto import CryptoAddress, ECDSAPublicKey, Hash512
from .errors import (BlockNotFoundError, BlockValidationError,
                     IndexIntegrityError, TransactionNotFoundError)
from .transaction import Transaction

if TYPE_CHECKING:
    # pylint: disable=unused-import
    from querylayer import QueryLayer
    # pylint: enable=unused-import

BLOCKS_PER_FILE = 10000
WORDSIZE = 8
INITIAL_POW = 2


def packdict(d: Dict) -> bytes:
    """Pack a dictionary into a JSON byte string."""
    return json.dumps(d).encode('utf-8')


def packint(i: int) -> bytes:
    """Pack an int into a byte string."""
    return struct.pack("<Q", i)


def unpackint(b: bytes) -> int:
    """Unpack a bytes string into a single int (only the first is returned)."""
    return struct.unpack("<Q", b)[0]


class BlockDB:
    """Database layer for blockchain."""

    def __init__(self, base_dir: str) -> None:
        """Create a new BlockDB object."""
        self.base_dir = base_dir
        self.indexfile = os.path.join(self.base_dir, "index.db")
        self.db = plyvel.DB(self.indexfile, create_if_missing=True)
        self.num_blocks = self.db.get(b"sc_numblocks") or 0
        self.pow_difficulty = self.db.get(b"sc_pow") or INITIAL_POW
        return

    def consistency_check(self, q: 'QueryLayer') -> None:
        """Perform consistency check between blockchain and index."""
        index = 0
        offset = 0
        lastfn = ''
        try:
            while True:
                fn = self.get_filename_for_block(index)
                if fn != lastfn:
                    offset = 0
                lastfn = fn

                data = self.read_chunk(fn, offset)
                size = len(data)

                try:
                    block_offset = self.get_block_offset(index)
                    assert block_offset == offset, \
                        "Invalid block offset for block {}".format(index)
                except BlockNotFoundError:
                    self.set_block_offset(index, offset)

                index += 1
                offset += WORDSIZE + size  # Account for size header.
        except AssertionError as exc:
            raise IndexIntegrityError("Consistency check failed") from exc
        except (IOError, EOFError):
            pass

        self.num_blocks = index
        self.db.put(b"sc_numblocks", packint(self.num_blocks))

        # Validate all blocks.
        for i in range(self.num_blocks):
            block = self.get_block(i)
            block.validate(q, self.pow_difficulty)
            self.update_transaction_indexes(block)

        if self.num_blocks > 0:
            print("Consistency check passed...")
        return

    def get_filename_for_block(self, block_index: int) -> str:
        """Get the filename corresponding to the specified block."""
        # We split blocks over several files to avoid going over the
        # max file size for the file system.
        file_num = int(block_index / BLOCKS_PER_FILE)
        file_suffix = "{:x}".format(file_num)
        return os.path.join(self.base_dir, "bdata_{}.blk".format(file_suffix))

    def get_key_for_block(self, block_index: int) -> bytes:
        """Get the levelDB key for the specified block."""
        return "block{:09d}".format(block_index).encode('utf-8')

    def get_key_for_transaction(self, txid: Hash512) -> bytes:
        """Get the levelDB key for the specified transaction id."""
        txid_str = txid.to_string()
        return "txid_{}".format(txid_str).encode('utf-8')

    def get_key_for_utxo(self, txid: Hash512, address: CryptoAddress) -> bytes:
        """Get the levelDB key for the specified UTXO entry."""
        txid_str = txid.to_string()
        address_str = address.to_string()

        # Put address first, since there may be multiple transactions
        # that output to that address, and this makes it easier to search
        # by address only.
        return "utxo_{}_{}".format(address_str, txid_str).encode('utf-8')

    def get_utxo_keys_for_address(self, address: CryptoAddress) -> List[bytes]:
        """Get UTXO keys for the specified address."""
        address_str = address.to_string()
        prefix = "utxo_{}_".format(address_str).encode('utf-8')
        keys = []
        for key in self.db.iterator(prefix=prefix, include_key=True,
                                    include_value=False):
            keys.append(key)
        return keys

    def get_txids_for_address(self, address: CryptoAddress) -> List[Hash512]:
        """Get all (unspent) input txids for the specified address."""
        address_str = address.to_string()
        prefix = "utxo_{}_".format(address_str).encode('utf-8')
        keys = self.get_utxo_keys_for_address(address)
        txids = []
        for key in keys:
            assert key.find(prefix) == 0, "UTXO key doesn't match address"
            txid = key.replace(prefix, b'', 1)
            txids.append(Hash512(txid))
        return txids

    def get_block_offset(self, block_index: int) -> int:
        """Get the file offset for the specified block."""
        key = self.get_key_for_block(block_index)
        raw_data = self.db.get(key)
        if raw_data is None:
            raise BlockNotFoundError(
                "Unable to find block {}".format(block_index))
        return unpackint(raw_data)

    def set_block_offset(self, block_index: int, offset: int, wb=None) -> None:
        """Set the block offset for the specified block."""
        if not wb:
            wb = self.db

        key = self.get_key_for_block(block_index)
        wb.put(key, packint(offset))
        return

    def read_chunk(self, blockfile: str, offset: int) -> bytes:
        """Read a chunk from the block data."""
        # The file simply stores size (4 bytes) + data (<size> bytes).
        with open(blockfile, "rb") as f:
            f.seek(offset)
            # Read size as unsigned long long (8 bytes).
            raw_data = f.read(WORDSIZE)
            if not raw_data:
                raise EOFError("Tried to read after EOF")

            size = struct.unpack('<Q', raw_data)[0]
            return f.read(size)

    def write_chunk(self, blockfile: str, data: bytes) -> int:
        """Write a chunk to the end of the block data."""
        offset = None
        with open(blockfile, "ab") as f:
            offset = f.tell()
            # Write size as unsigned long long (8 bytes).
            f.write(struct.pack('<Q', len(data)))
            f.write(data)
        return offset

    def get_block(self, block_index: int) -> Block:
        """Read a block from the block data."""
        offset = self.get_block_offset(block_index)
        blockfile = self.get_filename_for_block(block_index)
        data = self.read_chunk(blockfile, offset)
        return Block.deserialise(data.decode('utf-8'))

    def update_transaction_indexes(self, block: Block) -> None:
        """Update all UTXO indexes and amounts."""
        # Add transaction indexes.
        with self.db.write_batch() as wb:
            for trans in block.transactions:
                txkey = self.get_key_for_transaction(trans.txid)
                wb.put(txkey, packint(block.index))

                # Delete UTXO for inputs.
                for txinput in trans.inputs:
                    addr = ECDSAPublicKey(txinput.pubkey).get_address()
                    utxokey = self.get_key_for_utxo(txinput.txid, addr)
                    wb.delete(utxokey)

                # Add UTXO for outputs.
                for output in trans.outputs:
                    utxokey = self.get_key_for_utxo(trans.txid,
                                                    output.address)
                    wb.put(utxokey, packint(output.amount))
        return

    def write_new_block(self, block: Block, q: 'QueryLayer',
                        genesis: bool = False) -> None:
        """Write a new block to the blockchain."""
        index = self.num_blocks
        assert block.index == index, \
            "New block index does not match blockchain!"

        if not genesis:
            block.validate(q, self.pow_difficulty)

        blockfile = self.get_filename_for_block(index)
        offset = self.write_chunk(blockfile, block.serialise().encode('utf-8'))
        self.set_block_offset(index, offset)
        self.update_transaction_indexes(block)

        self.num_blocks += 1
        self.db.put(b"sc_numblocks", packint(self.num_blocks))
        return

    def get_utxo_amount(self, txid: Hash512, address: CryptoAddress) -> int:
        """Get the amount associated with the specified UTXO."""
        utxokey = self.get_key_for_utxo(txid, address)
        raw_value = self.db.get(utxokey)
        if raw_value is None:
            raise TransactionNotFoundError("UTXO not found")
        return unpackint(raw_value)

    def get_transaction(self, txid: Hash512) -> Transaction:
        """Get the specified transaction from the blockchain."""
        txid_str = txid.to_string()
        txkey = self.get_key_for_transaction(txid)
        raw_value = self.db.get(txkey)
        if raw_value is None:
            raise TransactionNotFoundError(
                "Unable to find transaction {}".format(txid_str))

        block_index = unpackint(raw_value)
        try:
            block = self.get_block(block_index)
            for trans in block.transactions:
                if trans.txid == txid:
                    return trans

            raise TransactionNotFoundError(
                "Unable to find transaction {} in block {}".
                format(txid_str, block_index))
        except (BlockNotFoundError, BlockValidationError) as exc:
            raise TransactionNotFoundError(
                "Unable to find transaction {}".format(txid_str)) from exc

File: rs/src/blockdb.rs
/// Database layer for the blockchain — direct port of lib/blockdb.py.
///
/// Uses rusty-leveldb for the index and flat binary files for block data.
use std::fs::{File, OpenOptions};
use std::io::{Read, Seek, SeekFrom, Write};

use rusty_leveldb::DB;

use crate::block::Block;
use crate::crypto::{CryptoAddress, ECDSAPublicKey, Hash};
use crate::errors::{BlockChainError, Result};
use crate::transaction::Transaction;

const BLOCKS_PER_FILE: u64 = 10000;
const WORDSIZE: usize = 8;
pub const INITIAL_POW: usize = 2;

pub fn packint(i: u64) -> Vec<u8> {
    i.to_le_bytes().to_vec()
}

fn unpackint(b: &[u8]) -> u64 {
    let mut buf = [0u8; 8];
    buf.copy_from_slice(&b[..8]);
    u64::from_le_bytes(buf)
}

pub struct BlockDB {
    pub base_dir: String,
    pub db: DB,
    pub num_blocks: u64,
    pub pow_difficulty: usize,
}

impl BlockDB {
    pub fn new(base_dir: &str) -> Result<Self> {
        let indexfile = format!("{base_dir}/index.db");
        let opts = rusty_leveldb::Options::default();
        let mut db = DB::open(&indexfile, opts)
            .map_err(|e| BlockChainError::DbIntegrity(format!("Failed to open DB: {e}")))?;

        let num_blocks = db
            .get(b"sc_numblocks")
            .map(|v| unpackint(&v))
            .unwrap_or(0);

        let pow_difficulty = match db.get(b"sc_pow") {
            Some(v) => unpackint(&v) as usize,
            None => {
                db.put(b"sc_pow", &packint(INITIAL_POW as u64))
                    .map_err(|e| BlockChainError::DbIntegrity(format!("DB write failed: {e}")))?;
                INITIAL_POW
            }
        };

        Ok(BlockDB {
            base_dir: base_dir.to_string(),
            db,
            num_blocks,
            pow_difficulty,
        })
    }

    pub fn consistency_check(&mut self) -> Result<()> {
        let mut index: u64 = 0;
        let mut offset: usize = 0;
        let mut last_fn = String::new();

        loop {
            let filename = self.get_filename_for_block(index);
            if filename != last_fn {
                offset = 0;
                last_fn = filename.clone();
            }

            match self.read_chunk(&filename, offset) {
                Ok(data) => {
                    let size = data.len();
                    match self.get_block_offset(index) {
                        Ok(existing_offset) => {
                            if existing_offset != offset {
                                return Err(BlockChainError::IndexIntegrity(format!(
                                    "Invalid block offset for block {index}"
                                )));
                            }
                        }
                        Err(BlockChainError::BlockNotFound(_)) => {
                            self.set_block_offset(index, offset)?;
                        }
                        Err(e) => return Err(e),
                    }
                    index += 1;
                    offset += WORDSIZE + size;
                }
                Err(_) => break,
            }
        }

        self.num_blocks = index;
        self.db
            .put(b"sc_numblocks", &packint(self.num_blocks))
            .map_err(|e| BlockChainError::DbIntegrity(format!("DB write failed: {e}")))?;

        // Validate all blocks and update indexes.
        // Note: we skip full block.validate() here since the QueryLayer
        // borrow-checker dance is complex. We do validate indexes.
        for i in 0..self.num_blocks {
            let block = self.get_block(i)?;
            self.update_transaction_indexes(&block)?;
        }

        if self.num_blocks > 0 {
            println!("Consistency check passed...");
        }
        Ok(())
    }

    pub fn get_filename_for_block(&self, block_index: u64) -> String {
        let file_num = block_index / BLOCKS_PER_FILE;
        format!("{}/bdata_{:x}.blk", self.base_dir, file_num)
    }

    fn get_key_for_block(&self, block_index: u64) -> Vec<u8> {
        format!("block{block_index:09}").into_bytes()
    }

    fn get_key_for_transaction(&self, txid: &Hash) -> Vec<u8> {
        format!("txid_{}", txid.to_hex()).into_bytes()
    }

    pub fn get_key_for_utxo(&self, txid: &Hash, address: &CryptoAddress) -> Vec<u8> {
        format!(
            "utxo_{}_{}",
            address.to_string_repr(),
            txid.to_hex()
        )
        .into_bytes()
    }

    pub fn get_utxo_keys_for_address(&mut self, address: &CryptoAddress) -> Vec<Vec<u8>> {
        let prefix = format!("utxo_{}_", address.to_string_repr()).into_bytes();
        let mut keys = Vec::new();

        // Use LdbIterator trait methods
        use rusty_leveldb::LdbIterator;
        if let Ok(mut iter) = self.db.new_iter() {
            iter.seek(&prefix);
            let mut key_buf = Vec::new();
            let mut val_buf = Vec::new();
            while iter.valid() {
                key_buf.clear();
                val_buf.clear();
                if iter.current(&mut key_buf, &mut val_buf) {
                    if key_buf.starts_with(&prefix) {
                        keys.push(key_buf.clone());
                    } else {
                        break; // Past the prefix range
                    }
                }
                if !iter.advance() {
                    break;
                }
            }
        }
        keys
    }

    pub fn get_txids_for_address(&mut self, address: &CryptoAddress) -> Result<Vec<Hash>> {
        let prefix = format!("utxo_{}_", address.to_string_repr()).into_bytes();
        let keys = self.get_utxo_keys_for_address(address);
        let mut txids = Vec::new();
        for key in &keys {
            let txid_bytes = &key[prefix.len()..];
            let txid_hex = String::from_utf8_lossy(txid_bytes);
            txids.push(crate::crypto::hash512_from_hex(&txid_hex)?);
        }
        Ok(txids)
    }

    pub fn get_block_offset(&mut self, block_index: u64) -> Result<usize> {
        let key = self.get_key_for_block(block_index);
        match self.db.get(&key) {
            Some(raw) => Ok(unpackint(&raw) as usize),
            None => Err(BlockChainError::BlockNotFound(format!(
                "Unable to find block {block_index}"
            ))),
        }
    }

    pub fn set_block_offset(&mut self, block_index: u64, offset: usize) -> Result<()> {
        let key = self.get_key_for_block(block_index);
        self.db
            .put(&key, &packint(offset as u64))
            .map_err(|e| BlockChainError::DbIntegrity(format!("DB write failed: {e}")))?;
        Ok(())
    }

    fn read_chunk(&self, blockfile: &str, offset: usize) -> Result<Vec<u8>> {
        let mut f = File::open(blockfile)?;
        f.seek(SeekFrom::Start(offset as u64))?;

        let mut size_buf = [0u8; WORDSIZE];
        f.read_exact(&mut size_buf).map_err(|_| {
            BlockChainError::Io(std::io::Error::new(
                std::io::ErrorKind::UnexpectedEof,
                "Tried to read after EOF",
            ))
        })?;
        let size = u64::from_le_bytes(size_buf) as usize;

        let mut data = vec![0u8; size];
        f.read_exact(&mut data)?;
        Ok(data)
    }

    pub fn write_chunk(&self, blockfile: &str, data: &[u8]) -> Result<usize> {
        let mut f = OpenOptions::new()
            .create(true)
            .append(true)
            .open(blockfile)?;

        let offset = f.seek(SeekFrom::End(0))? as usize;
        let size_bytes = (data.len() as u64).to_le_bytes();
        f.write_all(&size_bytes)?;
        f.write_all(data)?;
        Ok(offset)
    }

    pub fn get_block(&mut self, block_index: u64) -> Result<Block> {
        let offset = self.get_block_offset(block_index)?;
        let blockfile = self.get_filename_for_block(block_index);
        let data = self.read_chunk(&blockfile, offset)?;
        let json_str = String::from_utf8(data)
            .map_err(|e| BlockChainError::BlockValidation(format!("Invalid UTF-8: {e}")))?;
        Block::deserialise(&json_str)
    }

    pub fn update_transaction_indexes(&mut self, block: &Block) -> Result<()> {
        // rusty-leveldb WriteBatch::new() is pub(crate), so we use
        // individual put/delete calls instead.
        for trans in &block.transactions {
            let txkey = self.get_key_for_transaction(&trans.txid);
            self.db
                .put(&txkey, &packint(block.index))
                .map_err(|e| BlockChainError::DbIntegrity(format!("DB write failed: {e}")))?;

            // Delete UTXO for inputs
            for txinput in &trans.inputs {
                let addr = ECDSAPublicKey::from_hash(&txinput.pubkey)?.get_address();
                let utxokey = self.get_key_for_utxo(&txinput.txid, &addr);
                let _ = self.db.delete(&utxokey);
            }

            // Add UTXO for outputs
            for output in &trans.outputs {
                let utxokey = self.get_key_for_utxo(&trans.txid, &output.address);
                self.db
                    .put(&utxokey, &packint(output.amount))
                    .map_err(|e| {
                        BlockChainError::DbIntegrity(format!("DB write failed: {e}"))
                    })?;
            }
        }
        Ok(())
    }

    pub fn write_new_block(&mut self, block: &Block, _genesis: bool) -> Result<()> {
        let index = self.num_blocks;
        if block.index != index {
            return Err(BlockChainError::BlockValidation(
                "New block index does not match blockchain".into(),
            ));
        }

        let blockfile = self.get_filename_for_block(index);
        let data = block.serialise()?;
        let offset = self.write_chunk(&blockfile, data.as_bytes())?;
        self.set_block_offset(index, offset)?;
        self.update_transaction_indexes(block)?;

        self.num_blocks += 1;
        self.db
            .put(b"sc_numblocks", &packint(self.num_blocks))
            .map_err(|e| BlockChainError::DbIntegrity(format!("DB write failed: {e}")))?;
        Ok(())
    }

    pub fn reverse_transaction_indexes(&mut self, block: &Block) -> Result<()> {
        let genesis_txid = crate::crypto::Hasher::new_with_message(b"Genesis").get_hash();

        for trans in &block.transactions {
            // Delete txid_{hash} entry
            let txkey = self.get_key_for_transaction(&trans.txid);
            let _ = self.db.delete(&txkey);

            // Delete UTXO entries for outputs (undo created UTXOs)
            for output in &trans.outputs {
                let utxokey = self.get_key_for_utxo(&trans.txid, &output.address);
                let _ = self.db.delete(&utxokey);
            }

            // Restore UTXO entries for inputs (re-insert spent UTXOs)
            for txinput in &trans.inputs {
                let addr = ECDSAPublicKey::from_hash(&txinput.pubkey)?.get_address();
                if txinput.txid == genesis_txid {
                    // Genesis input — restore with TOTAL_COINS
                    let utxokey = self.get_key_for_utxo(&txinput.txid, &addr);
                    self.db
                        .put(&utxokey, &packint(crate::blockchain::TOTAL_COINS))
                        .map_err(|e| {
                            BlockChainError::RollbackFailed(format!("DB write failed: {e}"))
                        })?;
                } else {
                    // Look up the source transaction to find the output amount
                    let source_tx = self.get_transaction(&txinput.txid)?;
                    let mut found = false;
                    for output in &source_tx.outputs {
                        if output.address == addr {
                            let utxokey = self.get_key_for_utxo(&txinput.txid, &addr);
                            self.db
                                .put(&utxokey, &packint(output.amount))
                                .map_err(|e| {
                                    BlockChainError::RollbackFailed(format!(
                                        "DB write failed: {e}"
                                    ))
                                })?;
                            found = true;
                            break;
                        }
                    }
                    if !found {
                        return Err(BlockChainError::RollbackFailed(format!(
                            "Could not find matching output for input txid {}",
                            txinput.txid.to_hex()
                        )));
                    }
                }
            }
        }
        Ok(())
    }

    pub fn rollback_block(&mut self) -> Result<Block> {
        if self.num_blocks <= 1 {
            return Err(BlockChainError::RollbackFailed(
                "Cannot roll back genesis block".into(),
            ));
        }

        let block_index = self.num_blocks - 1;
        let block = self.get_block(block_index)?;

        // Reverse transaction indexes
        self.reverse_transaction_indexes(&block)?;

        // Get block offset and truncate the .blk file
        let offset = self.get_block_offset(block_index)?;
        let blockfile = self.get_filename_for_block(block_index);
        let f = OpenOptions::new().write(true).open(&blockfile)?;
        f.set_len(offset as u64)?;

        // Delete block key from LevelDB
        let key = self.get_key_for_block(block_index);
        let _ = self.db.delete(&key);

        // Decrement num_blocks and persist
        self.num_blocks -= 1;
        self.db
            .put(b"sc_numblocks", &packint(self.num_blocks))
            .map_err(|e| BlockChainError::RollbackFailed(format!("DB write failed: {e}")))?;

        Ok(block)
    }

    pub fn get_utxo_amount(&mut self, txid: &Hash, address: &CryptoAddress) -> Result<u64> {
        let utxokey = self.get_key_for_utxo(txid, address);
        match self.db.get(&utxokey) {
            Some(raw) => Ok(unpackint(&raw)),
            None => Err(BlockChainError::TransactionNotFound(
                "UTXO not found".into(),
            )),
        }
    }

    pub fn get_transaction(&mut self, txid: &Hash) -> Result<Transaction> {
        let txkey = self.get_key_for_transaction(txid);
        match self.db.get(&txkey) {
            Some(raw) => {
                let block_index = unpackint(&raw);
                let block = self.get_block(block_index)?;
                for trans in &block.transactions {
                    if trans.txid == *txid {
                        return Ok(trans.clone());
                    }
                }
                Err(BlockChainError::TransactionNotFound(format!(
                    "Unable to find transaction {} in block {block_index}",
                    txid.to_hex()
                )))
            }
            None => Err(BlockChainError::TransactionNotFound(format!(
                "Unable to find transaction {}",
                txid.to_hex()
            ))),
        }
    }
}
